test/functional/p2p_add_connections.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test add_outbound_p2p_connection test framework functionality"""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.p2p import P2PInterface
+from test_framework.util import assert_equal, disconnect_nodes
+
+
+class P2PAddConnections(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = False
+        self.num_nodes = 2
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        disconnect_nodes(self.nodes[0], 1)
+
+        self.log.info("Add 8 outbounds to node 0")
+        for i in range(8):
+            self.log.info("outbound: {}".format(i))
+            self.nodes[0].add_outbound_p2p_connection(P2PInterface())
+
+        self.log.info("Add 2 block-relay-only connections to node 0")
+        for i in range(2):
```

Comment:
When I change this from 2 to 3 I won't get an RPC error ("at capacity for connection type") as I would have expected, but an "address in use" OSError from the python asyncio. Would it be possible to have some tolerance here, so that the node behavior at maximum outbound connections can be tested?

Reply:
🤔 good question. I expected there to be some room because we allow 12 `MAX_NODES` and here we are using 10 of those ports, but indeed, I get the same issue. somehow that port is already being used, but I don't understand by what. its none of the other connections. I'll continue digging in. 

but will note that we can check the bounds of each type if they came first. eg. by adding 9 full-relay outbounds in previous loop, or adding 3 block-relay outbounds in the loop on line 33 (which come before the full relay for that node)

Reply:
2 bitcoin nodes, 8 full relay outbounds, and 2 blocks-only outbounds fills up the `MAX_NODES==12` ports

Reply:
@mzumsande do you think it would be useful to bump the port allocation, or does the workaround I proposed suffice for the use cases you're thinking about? I'm hesitant because of the potential of creating sporadic test framework issues, but I can look deeper into it if we have a good use case.

Reply:
Not sure, but I tend to think it is ok to keep it as it is. For none of the use cases I can think of, two bitcoin nodes are necessary.